### ISSUE TITLE
Unix File descriptor for memory map

### ISSUE THREAD

**Issue by dannyfriar** [dotnet/runtime#52854]
### Description

Context: I'm trying to create a non-persisted memory mapped file in C# and pass the value of the file handle to a child (Python) process for IPC.

I'm running on Ubuntu 18.04 with .NET Core 3.1.404 and trying to understand if there is a way to get the unix file descriptor from `mmap.SafeMemoryMappedFileHandle.DangerousGetHandle()`. Example code below:

```
using System;
using System.IO;
using System.IO.MemoryMappedFiles;

namespace myApp
    {
        class Program
        {
            static void Main(string[] args)
            {
                var mmap = MemoryMappedFile.CreateNew(null, 100, MemoryMappedFileAccess.ReadWrite, MemoryMappedFileOptions.None, HandleInheritability.Inheritable);
                Console.WriteLine($"Handle: {mmap.SafeMemoryMappedFileHandle.DangerousGetHandle()}");
    
                mmap.Dispose();
            }
        }
    }
```
    

```
>> dotnet run
Handle: 2
```

I checked the contents of `/proc/{pid}/fd` while the process was running and in my case it looks as if the actual assigned file descriptor for the non-persisted mmap was 30. 

Should this `mmap.SafeMemoryMappedFileHandle.DangerousGetHandle()` return the unix file descriptor? What does the return value from this function refer to when run on Linux?

When run on Windows, this works as expected and returns the windows file handle.

### Configuration

* Which version of .NET is the code running on? .NET Core 3.1.404
* What OS and version, and what distro if applicable? Ubuntu 18.04

**Comment by jkotas** [comment]
Memory mappings do not have real OS handles on Unix.

On Unix, `SafeMemoryMappedFileHandle.DangerousGetHandle` returns a fake handle that is generared by a global counter: https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs#L62-L63

You may want to look at `SafeFileHandle`. `SafeFileHandle` matches `/proc/{pid}/fd`.

**Comment by GPSnoopy** [comment]
@jkotas Is there a reason why we don't want to expose the MM handle on Unix?

**Comment by dannyfriar** [comment]
Thanks @jkotas.

I've been experimenting with using the file descriptor from `SafeMemoryMappedFileHandle._fileStream.SafeFileHandle.DangerousGetHandle()` (using reflection) in order to pass the file descriptor to a child Python process for sharing memory. I was wondering why this is not exposed as it seems that it can usefully shared with another process ([and it seems there's always a valid fd from either `shm_open` or a backing file in `/tmp`](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs#L167))

**Comment by jkotas** [comment]
What would that look like? As I have said, there is no equivalent of the Windows memory mapping handle on Unix.

**Comment by dannyfriar** [comment]
It looks as if there is always an associated file descriptor from either `shm_open` or a backing file in `/tmp`? Which would for example be accessed from the `SafeMemoryMappedFileHandle._fileStream`. I was curious as to why this is not exposed.

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs#L167-L171

**Comment by dannyfriar** [comment]
Thanks @jkotas. 

Even though the mapping can be backed by different OS primitives, they all yield a unix file descriptor. It still seems like it would be useful to expose this as an `IntPtr`, leaving it up to the user to know/decide how to handle it

**Comment by jkotas** [comment]
Sounds reasonable. Would you like to create a PR with the proposed change?

**Comment by jkotas** [comment]
I assume that we would just reuse the existing SafeHandle API for that. If this change was to introduce new APIs, it would have to go through https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

**Comment by dannyfriar** [comment]
Sure, I'm happy to create a PR for this. I think we can reuse the existing SafeHandle API.

**Comment by dannyfriar** [comment]
@jkotas  Took a bit of a look at this, my original thought was to just set the `SafeMemoryMappedFileHandle.handle` to the `fileStream` handle and remove the existing counter, _roughly_ [these lines](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs#L62-L64)  would become something like:

```
IntPtr handle = fileStream != null ? fileStream.SafeFileHandle.DangerousGetHandle() : new IntPtr(-1);
SetHandle(handle);
```

However, there is a case on Unix with read-only maps where nothing backs it. What value do you think should be returned as the handle in this case? In a way, I think `-1` should be returned since the handle _is_ invalid but this doesn't fit well with the existing implementation (`SafeHandleZeroOrMinusOneIsInvalid` etc).

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs#L97-L106

**Comment by jkotas** [comment]
> IntPtr handle = fileStream != null ? fileStream.SafeFileHandle.DangerousGetHandle() : new IntPtr(-1);

I do not think that this would work well. Each SafeHandle instance has to own the underlaying handle. Otherwise, we can end up with safe handles that point to closed handles, etc.

I think this would have to create a duplicate of the file stream handle.

> I think -1 should be returned since the handle is invalid but this doesn't fit well with the existing implementation (SafeHandleZeroOrMinusOneIsInvalid etc).

Some of the natural choices would be `-2` or `IntPtr.MaxValue`.
